server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -78,31 +84,53 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<ReplicaShardRefreshRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
+        try (var listeners = new RefCountingListener(listener.map(v -> new PrimaryResult<>(shardRequest, new ReplicationResponse())))) {
             var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            var shardRefreshRequest = new ReplicaShardRefreshRequest(
-                primary.shardId(),
-                shardRequest.getParentTask(),
-                refreshResult.generation()
-            );
-            return new PrimaryResult<>(shardRefreshRequest, new ReplicationResponse());
-        });
+
+            // Forward the request to all nodes that hold unpromotable replica shards
+            final ClusterState clusterState = clusterService.state();
+            final Task parentTaskId = taskManager.getTask(shardRequest.getParentTask().getId());
+            clusterState.routingTable()
+                .shardRoutingTable(shardRequest.shardId())
+                .assignedShards()
+                .stream()
+                .filter(Predicate.not(ShardRouting::isPromotableToPrimary))
+                .map(ShardRouting::currentNodeId)
+                .collect(Collectors.toUnmodifiableSet())
```

Comment:
Do we need this intermediate collection? I think it could be replaced with `distinct`

Reply:
I thought we dropped this. We don't even need `distinct()`, we're looking at node IDs for the copies of a single shard which are necessarily distinct anyway.

Reply:
I think we dropped it from another place, but not here. I can handle this in another PR.